README.md
# shmfile
Small shared library to use shared memory as a FILE\* , with interprocess communication in mind.



Basic Design
============

**Overview**

shmfile is desgined in an "owner/guest" format.

One process is the *owner* [FSHM\_OWMER] of a shared mapping (given by a string name). It creates the mapping, and can destroy it.


Other processes then can map that same string name as *guest* [FSHM\_GUEST].


The guest only needs to know the unique name that the owner picked to have access to the data.


**Permissions**

Permissions for who can map the same shm as FSHM\_GUEST and what they can do with the shmfile are specified by the octal: "mode".

These follow the same rules as chmod.

For example, same user only and read-only would be 0400. Same user r/w, all other users r/o, 0644


You specify the mode when the *owner* creates the shmfile, which will later determine whomelse on the system may have access.

The permissions are specified by the FSHM\_OWNER when creating the shmfile.

If fshm\_open is used instead of fshm\_guest\_open to map an existing shmfile stream as FSHM\_GUEST, the mode argument is ignored.


**Usage**


The *fshm\_open* is the core function that returns a FILE\* object, with a valid integer file descriptor [ fd ]. It is passed either FSHM\_OWNER or FSHM\_GUEST in the "flags" argument to define the operation.

The owner may also create an shmfile stream by using the helper function, *fshm\_create*, and a guest may also map an existing shmfile stream by using the helper function, *fshm\_guest\_open*.

You can read/write the data using either the buffered I/O functions (like fread, fwrite, fseek) and flushing with fflush, OR

the unbuffered I/O functions (like read, write, lseek) by using the fd returned by fileno( fileObj ) where fileObj is the FILE\* returned by one of the aforementioned functions. 


The size is completely automatic, you do not need to allocate an underlying buffer -- any data written to it and flushed will
expand the buffer automatically.


There is no built-in locking, so it's suggested if you are going to have multiple processes writing to the stream, 

that you either use fixed-size sections for each process, or implement your own locking.


Public Functions
----------------

The following functions make up the public API of shmfile.


**fshm\_create** - Create a new shmfile stream (convenience/helper function for fshm\_open with fshm\_flags as FSHM\_OWNER)

FILE\* fshm\_create(const char \*name, mode\_t mode)

	/**
	 * fshm_create - Create a new shmfile (as FSHM_OWNER)
	 *                 with a given mode.
	 *
	 *
	 *  name - A unique name which corrosponds to this stream.
	 *          Guests will use this name as a reference to map
	 *           this shmfile stream.
	 *
	 *  mode - An octal representing the "mode" for guest mappings.
	 *           Mode has same meaning as with chmod.
	 *         This mode will be used to determine how FSHM_GUEST
	 *           mappings will be able to access this stream.
	 *
	 *         See the 'fshm_open' function for more information
	 *          on mode.
	 *
	 *
	 *  RETURN VALUE - 
	 *                 A pointer to a FILE structure which may be used
	 *                  with standard io functions (e.x. fread, fprintf, fseek)
	 *
	 *                 On error, this function will return NULL and `errno' will be set
	 *
	 *  NOTES -
	 *                 * This is the same as calling fshm_open(name, mode, FSHM_OWNER)
	 *
	 *                 * See fshm_open for more info
	 */


**fshm\_guest\_open** - Opens an existing shmfile stream (convenience/helper function for fshm\_open with fshm\_flags as FSHM\_GUEST)


FILE \*fshm\_guest\_open(const char \*name);

	/**
	 * fshm_guest_open - Maps an existing shmfile stream (as FSHM_GUEST)
	 *                    using the given name for reference.
	 *
	 *
	 *  name - The name associated with an already-created fshmfile stream
	 *
	 *
	 *  RETURN VALUE -
	 *                 A pointer to a FILE structure which may be used
	 *                  with standard io functions (e.x. fread, fprintf, fseek)
	 *
	 *                 On error, this function will return NULL and `errno' will be set
	 *
	 *  NOTES -
	 *                 * This is the same as calling fshm_open(name, 0, FSHM_GUEST)
	 *
	 *                 * See fshm_open for more info
	 */

**fshm\_open** - Opens a shmfile stream

FILE\* fshm\_open(const char \*name, mode\_t mode, int fshm\_flags);


This is the "core" function, and can be used both to create a new shared memory stream or open an existing stream. The FILE pointer returned allows you to use all the other I/O systems that come standard with libc.


	/**
	 * fshm_open - Open a shared memory stream, either creating a new
     *   stream [FSHM_OWNER] or mapping an existing stream [FSHM_GUEST],
     *   and returning a FILE* object with an associated int fd.
	 *   This will work with both FILE* and int filedes functions
	 *    (e.x.  fwrite and write, fread and read, etc).
	 *
	 *  name - A unique name which corrosponds to this stream.
	 *            If flags contain FSHM_OWNER, we attempt to create
	 *             a stream using this name.
	 *            Guests who wish to map this shmfile will provide
	 *             this same name to reference which shmfile stream
	 *             to map.
	 *
	 *  mode - An octal representing the "mode" for this file stream.
	 *           Mode has same meaning as with chmod.
	 *
	 *         This mode will be used to determine how FSHM_GUEST
	 *           mappings will be able to access this stream.
	 *
	 *         It will also define the permissions (read/write/execute) for the
	 *           owner / group-member / other
	 *         relations on mappings.
	 *
	 *
	 *         This field only has meaning when creating the shmfile ( FSHM_OWNER ).
	 *          For an FSHM_GUEST open of an existing shmfile stream, this can be
	 *            set to 0. Any other value will be ignored when flags contains FSHM_GUEST.
	 *
	 *         Columns:
	 *
	 *          - First column from left is always 0 for now (no sticky/setguid bit meaning)
	 *
	 *          - Second column from left is permissions for owner (user who creates shmfile)
	 *
	 *          - Third column from the left is permissions for users who are a member of
	 *            the group associated with this stream. 
	 *            This could be primary group of creating user (FSHM_OWNER), or a different
	 *            group (set by fshm_chgrp).
	 *            
	 *          - Fourth column from the left are permissions for any user which
	 *            does NOT fall into any categories listed above
	 *
	 *         Values:
	 *
	 *          For each column, the number is derived by starting with 0
	 *            and following these rules:
	 *
	 *           - Add 4 to allow FSHM_GUEST mapping and read for represented set
	 *           - Add 2 to allow writing to the mapping
	 *           - Add 1 to allow execution
	 *
	 *
	 *  Examples - Some examples of various #mode[s] and their meaning
	 *
	 *        Allow current user only to map as guest, read/write:
	 *           0600
	 *
	 *        Allow current user only to map as guest, read-only:
	 *           0400
	 *
	 *        Allow current user to map as guest, read and write,
	 *          allow a different user belonging to same primary group
	 *          to map as guest read-only:
	 *           0640
	 *
	 *        Allow anybody on system to map read/write:
	 *           0777
	 *
	 *
	 *
	 *  fshm_flags - A list of flags OR'd together
	 *
	 *     FSHM_OWNER  - Specifies that this call will create the stream
	 *                    specified by #name . When the owner closes the stream,
	 *                    it will be destroyed.
	 *
	 *     FSHM_GUEST  - Specifies that this call will open an existing stream
	 *                    specified by #name . An FSHM_OWNER must have created
	 *                    this stream already, or the call will fail.
	 *                    Closing the stream as a guest will not destroy the file.
	 *
	 *     FSHM_PERSIST - Do NOT remove this stream when the FSHM_OWNER closes it.
	 *                     fshm_force_destroy must be used to destroy the stream
	 *                     when this flag is provided.
	 *
	 *  Return:
	 *
	 *     A FILE* object on success, otherwise NULL and errno is set.
	 *
	 *  NOTES:
	 *
	 *     The idea here is to simply allow interprocess I/O that exists
	 *      only in memory. One process creates the mapping ( FSHM_OWNER )
	 *      and several other processes can inherit that mapping,
	 *      and share data between them.
	 *
	 *     There is a real FD associated with this FILE (unlike, open_memstream, for example).
	 *
	 *     You may choose to use either the buffered variants ( like fwrite, fread),
	 *       or the unbuffered variants (like write, read).
	 *
	 *     Keep in mind that if you use the buffered functions, you must call fflush to
	 *       sync the changes.
	 *
	 *
	 *     You can use this object anywhere a FILE* is allowed, or anywhere an fd is allowed
	 *      by calling fileno(X) where X is the FILE* returned by this function.
	 *
	 */

**fshm\_chgrp** - Change the group associated with an shmfile stream

int fshm\_chgrp(FILE \*fshm\_file, gid\_t group)

	/**
	 * fshm_chgrp - Change the group associated with an existing shmfile stream
	 *
	 *      You must be the the same user as the FSHM_OWNER of the given stream,
	 *        or root in order to change the gid assigned to the shmfile.
	 *
	 *      The shmfile created by fshm_open is assigned the gid
	 *        matching the primary group of the creating user.
	 *
	 *      The purpose is to allow guests in a different group access rights to
	 *        this shmfile ( based on the group bits in the mode set with fshm_open ),
	 *        but not just open up access for everybody ( using the "other" bits in mode ).
	 *
	 *      As non-root, you may change the group to any groups to which you are a member.
	 *        This includes your primary gid as well as all supplementry groups to which you belong.
	 *        As root, you may change the shmfile's group to any group existing on this system.
	 *
	 *
	 *  fshm_file - A FILE* object as returned by fshm_open.
	 *                 You must be the FSHM_OWNER of this shmfile to change permissions.
	 *
	 *
	 *  group     - new gid to replace current group on this shmfile.
	 *
	 *                  If you are not root, this must be either your primary gid
	 *                    or a group contained in your list of supplementry groups.
	 *
	 *
	 *  RETURN VALUE -
	 *                   0:  Success
	 *                  -1:  Failure (and errno will be set)
	 *
	 *   See man 2 chown for possible error conditions and values of errno
	 * 
	 */


**fshm\_chmod** - Change the mode (permissions) on this shmfile

int fshm\_chmod(FILE \*fshm\_file, mode\_t mode);

	/**
	 * fshm_chmod - Change the mode on a shmfile
	 *
	 *   fshm_file - A FILE* object representing an shmfile stream
	 *
	 *   mode - An octal of permissions. See "chmod" for meaning.
	 *
	 *  RETURN VALUE -
	 *                   0:  Success
	 *                  -1:  Failure (and errno will be set)
	 *
	 *   See man 2 chmod for possible error conditions and values of errno
	 */ 


**fshm\_force\_destroy** - Forcibly destroys a stream, as a recovery tool.

Use this function if the FSHM\_OWNER of a shmfile stream dies before calling fclose on the stream.

int  fshm\_force\_destroy(const char \*name);

	/**
	 * fshm_force_destroy - Forcibly destroy the shared memory region associated
	 *                       with #name.
	 *
	 *   Normal operation does not require this, however, if a process opens
	 *     a fshm stream as FSHM_OWNER, and does not close it before terminating,
	 *     the stream will maintain stuck open without an owner.
	 *
	 *
	 *    name - The name associated with this stream
	 *
	 *  NOTES - 
	 *
	 *   The owner should test for this, and forcibly destroy if required.
	 *
	 *   Like so:
	 *
	 *   FILE *fObj;
	 *
	 *   errno = 0;
	 *   fObj = fshm_open("/example1", 0775, FSHM_OWNER);
	 *
	 *   if ( !fObj && errno == EEXIST )
	 *   {
	 *      fshm_force_destroy("/example1");
	 *      fObj = fshm_open("/example1", 0775, FSHM_OWNER);
	 *   }
	 *
	 *
	 *  RETURN:
	 *    0 on success, -1 on error.
	 */


**shmfile\_get\_version** - Gets version info on libshmfile

void shmfile\_get\_version(unsigned char \*major, unsigned char \*minor, unsigned char \*patchlevel, const char \*\*extra);

	/**
	 * shmfile_get_version - Get the version info on this version of shmfile
	 *
	 *    Sets the value of *major, *minor, *patchlevel, and *extra to the numeric version numbers
	 */


**Defined Constants**

*FSHM\_OWNER* - Used as a flag with fshm\_open to denote that the current process and user will create and own the shmfile stream.

*FSHM\_GUEST* - Used as a flag with fshm\_open to denote that we want to map an existing shmfile stream.

*FSHM\_PERSIST* - Used as a flag with fshm\_open to denote that we should not automatically destroy the stream when the FSHM\_OWNER closes it. If the owner provides this flag then the stream must be explicitly destroyed by calling fshm\_force\_destroy


Examples
--------

The source code contains the following examples/tests of libshmfile which can be used as references:

**examples/owner** ( from examples/owner.c )  is an example that creates a stream ( as FSHM\_OWNER ),
 writes some data to it, and leaves it open for a period of time.

**examples/guest** ( from examples/guest.c )  is an example that connects to an existing stream ( as FSHM\_GUEST ),
 reads data from it, and prints it to stdout.


These examples test using a struct as data which is given initial values by "owner", and is updated each time "guest" is run.

The examples show both buffered I/O (fwrite and fflush), as well as unbuffered I/O (read and write).

Optionally, the fshm\_chgrp method will be used if you uncomment and set the SET\_GID\_MACRO at the top of examples/owner.c



Changelog
1.1.0 - Nov 13 2018

* Add FSHM_PERSIST flag as a possible option to be OR'd into fshm_flags on fshm_open.
    - When present, this will cause the shmfile to not be deleted when the FSHM_OWNER closes it.
    - The stream must be explicitly destroyed by fshm_force_destroy

* Add man page, available under: fshm_open, fshm_create, fshm_guest_open, fshm_force_destroy

* Cleanup Makefile a bit


1.0.0 - Jul 24 2018

* Makefile: updates, bugfixes, expansion, cleanup

 - Makefile: Remove *.o files as part of clean

 - Fix the static build of the lib (it was building a dynamic library named .a prior... whoops!)

 - Add a target, "install_static" which works the same way as "install" but only installs the archive (libshmfile.a)
 - Add a target, "static_debug" which recompiles the static lib using debug flags

 - Split out libshmfile.o into its own unit

* Cleanup and separate examples from base code

 - Create a new directory "examples"
 - Move 'owner.c' and 'guest.c' and 'owner_guest_private.h' into this 'examples' directory

 - Remove "examples" from the "all" target into its own target "examples" as the example code is not required to build libshmfile.so or libshmfile.a
  
  * Add "examples" target which will compile the example code
  * Add "examples_debug" target which will compile the example code with debug CFLAGS

* Ensure that the "mode" provided to fshm_open is actually applied to the file. Now you can properly define rights (read/write) to owning user, owning group (now modifable by fshm_chgrp), and all other.

* Implement a fshm_chmod function which can modify permissions afer a shmfile has been created

* Implement a fshm_chgrp function which changes the group associated with the shmfile ( i.e. members of this group get max permissions as specified by the "group" bit in the "mode" parameter )

* Add a new function, fshm_create(char *name, mode_t mode) for creating a new shmfile, which is a shorthand for fshm_open(name, mode, FSHM_OWNER)

* Add a new function, fshm_guest_open(char *name) for mapping an existing shmfile, which is a shorthand for fshm_open(name, 0, FSHM_GUEST)

* Update shmfile_get_version to use "unsigned char" instead of "char" for the version info (major, minor, patchlevel) as the version components will never be signed (negative)


shmfile.c
/**
 * Copyright (c) 2017, 2018 Timothy Savannah
 *   All Rights Reserved, Licensed under terms of Lesser GNU Public License version 2.1
 *
 * shmfile.c - Definitions and private attributes for shmfile
 *
 */


#define _GNU_SOURCE

#include "shmfile.h"


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>

#ifdef __GNUC__

    #define __visibility_hidden  __attribute__( (visibility ("hidden")) )
    #define likely(x)    __builtin_expect(!!(x),1)
    #define unlikely(x)  __builtin_expect(!!(x),0)

#else

    #define __visibility_hidden
    #define likely(x)   ( !!(x) )
    #define unlikely(x) ( !!(x) )

#endif


#ifndef FSHM_QUIET

#define printerr(_x, ...) fprintf(stderr, _x)

#else

#define printerr(_x, ...)

#endif

#define FSHM_COOKIE(_cookie) ( (struct fshm_cookie*) _cookie)


#define IS_FSHM_OWNER(_cookie) !!( FSHM_COOKIE(_cookie)->fshm_flags & FSHM_OWNER )
#define IS_FSHM_GUEST(_cookie) !!( FSHM_COOKIE(_cookie)->fshm_flags & FSHM_GUEST )

#define IS_FSHM_PERSISTENT(_cookie) !!( FSHM_COOKIE(_cookie)->fshm_flags & FSHM_PERSIST )

#define IS_FSHM_FLAGS_OWNER(_fshm_flags) !!( _fshm_flags & FSHM_OWNER )
#define IS_FSHM_FLAGS_GUEST(_fshm_flags) !!( _fshm_flags & FSHM_GUEST )

#define IS_FSHM_FLAGS_PERSIST(_fshm_flags) !!( _fshm_flags & FSHM_PERSIST )


static volatile const char *copyright = "Copyright (c) 2017, 2018 Timothy Savannah   All Rights Reserved,  licensed under terms of Lesser GNU Public License version 2.1";

static volatile const char *product_name = "shmfile";
static volatile const char *version   = "Version 1.1.0";

typedef struct {
    unsigned char major;
    unsigned char minor;
    unsigned char patchlevel;
    const char *extra;
}_shmfile_version_info_t;

static volatile _shmfile_version_info_t _shmfile_version_info = { (unsigned char)1, (unsigned char)1, (unsigned char)0, "" };


void shmfile_get_version(unsigned char *major, unsigned char *minor, unsigned char *patchlevel, const char **extra)
{
    *major = _shmfile_version_info.major;
    *minor = _shmfile_version_info.minor;
    *patchlevel = _shmfile_version_info.patchlevel;
    *extra = _shmfile_version_info.extra;
}

void _fshm_printCopyrightVersion(void)
{
    printf("%s %s\n%s\n\n", product_name, version, copyright);
}

__visibility_hidden struct fshm_cookie {

    int fd;
    char *name;
    int fshm_flags;

};

__visibility_hidden ssize_t _fshm_read(void *cookie, char *buf, size_t size)
{
    struct fshm_cookie *fshmCookie;
    ssize_t ret;

    fshmCookie = FSHM_COOKIE(cookie);


    ret = read(fshmCookie->fd, buf, size);

    return ret;
}

__visibility_hidden ssize_t _fshm_write(void *cookie, const char *buf, size_t size)
{
    struct fshm_cookie *fshmCookie;
    ssize_t ret;

    fshmCookie = FSHM_COOKIE(cookie);

    ret = write(fshmCookie->fd, buf, size);

    return ret;
}

__visibility_hidden int _fshm_seek(void *cookie, off_t *offset, int whence)
{
    struct fshm_cookie *fshmCookie;
    int ret;
    off_t newOffset;

    fshmCookie = FSHM_COOKIE(cookie);

    newOffset = lseek(fshmCookie->fd, *offset, whence);

    if ( likely( newOffset >= 0 ) )
    {
        *offset = newOffset;
        ret = 0;
    }
    else
    {
        ret = -1;
    }

    return ret;
}

__visibility_hidden int _fshm_close(void *cookie)
{
    struct fshm_cookie *fshmCookie;
    int ret = 0;

    fshmCookie = FSHM_COOKIE(cookie);

    if ( IS_FSHM_OWNER(fshmCookie) && !IS_FSHM_PERSISTENT(cookie) )
    {
        ret = shm_unlink(fshmCookie->name);
    }

    free(fshmCookie->name);
    free(fshmCookie);

    return ret;
}

FILE* fshm_open(const char *name, mode_t mode, int fshm_flags)
{
    int fd;
    int oflag;

    char *nameCpy;
    size_t nameLen;
    const char *modeStr = NULL;

    struct fshm_cookie *cookie;
    cookie_io_functions_t fshmIoFuncs;

    FILE *fObj;


    if ( unlikely( ! ( fshm_flags & (FSHM_OWNER | FSHM_GUEST) ) ) )
    {
        printerr("Error: fshm_open called but flags specified neither FSHM_OWNER nor FSHM_GUEST.\n");
        errno = EINVAL;
        return NULL;
    }

    if ( unlikely( !name || name[0] == '\0' ) )
    {
        printerr("Error: fshm_open called with NULL/empty \"name\" parameter.\n");
        errno = EINVAL;
        return NULL;
    }

    oflag = O_RDWR;

    if ( fshm_flags & FSHM_OWNER )
    {
        if ( unlikely( fshm_flags & FSHM_GUEST ) )
        {
            printerr("Error: fshm_open called with both FSHM_OWNER and FSHM_GUEST\n");
            errno = EINVAL;
            return NULL;
        }
        oflag |= O_CREAT;
        oflag |= O_EXCL;
    }

    fd = shm_open(name, oflag, mode);

    if ( unlikely( fd < 0 ) )
    {
        if ( fshm_flags & FSHM_GUEST && errno == EACCES)
        {
            /*
             * Guest read and write denied,
             * Try read only
             * */
            oflag = O_RDONLY;
            fd = shm_open(name, oflag, mode);
            if ( fd >= 0 )
            {
                modeStr = "r";
                goto _created_ok;
            }
        }
        return NULL;
    }
_created_ok:

    cookie = malloc( sizeof(struct fshm_cookie) );
    if ( unlikely( !cookie ) )
    {
        errno = ENOMEM;
        return NULL;
    }

    nameLen = strlen(name);
    nameCpy = malloc(nameLen + 1);

    strcpy(nameCpy, name);

    cookie->fshm_flags = fshm_flags;
    cookie->fd = fd;
    cookie->name = nameCpy;

    fshmIoFuncs.read = _fshm_read;
    fshmIoFuncs.write = _fshm_write;
    fshmIoFuncs.seek = _fshm_seek;
    fshmIoFuncs.close = _fshm_close;


    if ( fshm_flags & FSHM_OWNER )
    {
        modeStr = "w+";
    }
    else if ( likely(modeStr == NULL) )
    {
        modeStr = "r+";
    }

    fObj = fopencookie(cookie, modeStr, fshmIoFuncs);
    if ( unlikely( ! fObj ) )
    {
        _fshm_close(cookie);
        return NULL;
    }

    /* Set the _fileno field so fileno() and then non-f* functions work  */
    fObj->_fileno = fd;

    if ( fshm_flags & FSHM_OWNER )
    {
        /* Set the mode permissions on this shmfile */
        fchmod(fd, mode);
    }

    return fObj;
}

FILE* fshm_create(const char *name, mode_t mode)
{
    return fshm_open(name, mode, FSHM_OWNER);
}


FILE *fshm_guest_open(const char *name)
{
    return fshm_open(name, 0, FSHM_GUEST);
}


int fshm_chgrp(FILE *fshm_file, gid_t group)
{
    int fd;
    uid_t ownerUid;
    struct stat statBuf;

    fd = fileno( fshm_file );
    /* Check if fshm_file was a valid stream */
    if ( unlikely( fd < 0 ) )
    {
        return -1;
    }

    /* Gather current uid on file */
    if ( unlikely( fstat( fd, &statBuf ) != 0 ) )
    {
        return -1;
    }

    ownerUid = statBuf.st_uid;

    /* Check if we are trying to change group to the current group (NOP) */
    if ( unlikely ( statBuf.st_gid == group ) )
    {
        return 0;
    }

    /* Perform the actual chown call. */
    if ( unlikely( fchown(fd, ownerUid, group) != 0 ) )
    {
        return -1;
    }

    return 0;
}

int fshm_chmod(FILE *fshm_file, mode_t mode)
{
    int fd;

    fd = fileno( fshm_file );

    /* Check if fshm_file was a valid stream */
    if ( unlikely( fd < 0 ) )
    {
        errno = ENOENT;
        return -1;
    }

    /* Perform the actual chmod call */
    if ( unlikely( fchmod(fd, mode) != 0 ) )
    {
        return -1;
    }

    return 0;
}



int fshm_force_destroy(const char *name)
{
    return shm_unlink(name);
}

/* vim: set ts=4 sw=4 st=4 expandtab : */


shmfile.h
#ifndef _SHM_FILE_H
#define _SHM_FILE_H
/**
 * Copyright (c) 2017, 2018 Timothy Savannah
 *   All Rights Reserved, Licensed under terms of Lesser GNU Public License version 2.1
 *
 * shmfile.h - Header for shmfile
 *
 *    https://github.com/kata198/shmfile
 *
 * Allows mapping shared-memory in an owner/guest fashion, usable through standard
 *   file I/O functions
 *
 */


/*#if defined(_FILE_OFFSET_BITS) && _FILE_OFFSET_BITS != 64
#error "_FILE_OFFSET_BITS must be set to 64. Currently defined to something else."
#endif

#define _FILE_OFFSET_BITS 64

#define _GNU_SOURCE
*/

#include <stdio.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>

#if defined (__cplusplus)
extern "C" {
#endif


/***************************
 *  Options for fshm_flags
 **************************/

#define FSHM_OWNER 1  /* FSHM_OWNER - This process will create and manage this stream. */
#define FSHM_GUEST 2  /* FSHM_GUEST - This process will inherit an existing stream */

#define FSHM_PERSIST 4 /* FSHM_PERSIST - Do NOT remove this stream when the FSHM_OWNER closes it.
                        *                 fshm_force_destroy must be used to destroy the stream
                        *                 when this flag is provided.
                        */

/*******************
 * PUBLIC API
 *************/

/**
 * fshm_create - Create a new shmfile (as FSHM_OWNER)
 *                 with a given mode.
 *
 *
 *  name - A unique name which corrosponds to this stream.
 *          Guests will use this name as a reference to map
 *           this shmfile stream.
 *
 *  mode - An octal representing the "mode" for guest mappings.
 *           Mode has same meaning as with chmod.
 *         This mode will be used to determine how FSHM_GUEST
 *           mappings will be able to access this stream.
 *
 *         See the 'fshm_open' function for more information
 *          on mode.
 *
 *
 *  RETURN VALUE - 
 *                 A pointer to a FILE structure which may be used
 *                  with standard io functions (e.x. fread, fprintf, fseek)
 *
 *                 On error, this function will return NULL and `errno' will be set
 *
 *  NOTES -
 *                 * This is the same as calling fshm_open(name, mode, FSHM_OWNER)
 *
 *                 * See fshm_open for more info
 */
FILE* fshm_create(const char *name, mode_t mode);


/**
 * fshm_guest_open - Maps an existing shmfile stream (as FSHM_GUEST)
 *                    using the given name for reference.
 *
 *
 *  name - The name associated with an already-created fshmfile stream
 *
 *
 *  RETURN VALUE -
 *                 A pointer to a FILE structure which may be used
 *                  with standard io functions (e.x. fread, fprintf, fseek)
 *
 *                 On error, this function will return NULL and `errno' will be set
 *
 *  NOTES -
 *                 * This is the same as calling fshm_open(name, 0, FSHM_GUEST)
 *
 *                 * See fshm_open for more info
 */
FILE *fshm_guest_open(const char *name);


/**
 * fshm_open - Open a shared memory stream, either creating a new
 *   stream [FSHM_OWNER] or mapping an existing stream [FSHM_GUEST],
 *   and returning a FILE* object with an associated int fd.
 *   This will work with both FILE* and int filedes functions
 *    (e.x.  fwrite and write, fread and read, etc).
 *
 *
 *  name - A unique name which corrosponds to this stream.
 *            If flags contain FSHM_OWNER, we attempt to create
 *             a stream using this name.
 *            Guests who wish to map this shmfile will provide
 *             this same name to reference which shmfile stream
 *             to map.
 *
 *  mode - An octal representing the "mode" for this file stream.
 *           Mode has same meaning as with chmod.
 *
 *         This mode will be used to determine how FSHM_GUEST
 *           mappings will be able to access this stream.
 *
 *         It will also define the permissions (read/write/execute) for the
 *           owner / group-member / other
 *         relations on mappings.
 *
 *         This field only has meaning when creating the shmfile ( FSHM_OWNER ).
 *          For an FSHM_GUEST open of an existing shmfile stream, this can be
 *            set to 0. Any other value will be ignored when flags contains FSHM_GUEST.
 *
 *         Columns:
 *
 *          - First column from left is always 0 for now (no sticky/setguid bit meaning)
 *
 *          - Second column from left is permissions for owner (user who creates shmfile)
 *
 *          - Third column from the left is permissions for users who are a member of
 *            the group associated with this stream. 
 *            This could be primary group of creating user (FSHM_OWNER), or a different
 *            group (set by fshm_chgrp).
 *            
 *          - Fourth column from the left are permissions for any user which
 *            does NOT fall into any categories listed above
 *
 *         Values:
 *
 *          For each column, the number is derived by starting with 0
 *            and following these rules:
 *
 *           - Add 4 to allow FSHM_GUEST mapping and read for represented set
 *           - Add 2 to allow writing to the mapping
 *           - Add 1 to allow execution
 *
 *
 *  Examples - Some examples of various #mode[s] and their meaning
 *
 *        Allow current user only to map as guest, read/write:
 *           0600
 *
 *        Allow current user only to map as guest, read-only:
 *           0400
 *
 *        Allow current user to map as guest, read and write,
 *          allow a different user belonging to same primary group
 *          to map as guest read-only:
 *           0640
 *
 *        Allow anybody on system to map read/write:
 *           0777
 *
 *
 *
 *  fshm_flags - A list of flags OR'd together
 *
 *     FSHM_OWNER  - Specifies that this call will create the stream
 *                    specified by #name . When the owner closes the stream,
 *                    it will be destroyed.
 *
 *     FSHM_GUEST  - Specifies that this call will open an existing stream
 *                    specified by #name . An FSHM_OWNER must have created
 *                    this stream already, or the call will fail.
 *                    Closing the stream as a guest will not destroy the file.
 *
 *     FSHM_PERSIST - Do NOT remove this stream when the FSHM_OWNER closes it.
 *                     fshm_force_destroy must be used to destroy the stream
 *                     when this flag is provided.
 *                       
 *
 *  Return:
 *
 *     A FILE* object on success, otherwise NULL and errno is set.
 *
 *  NOTES:
 *
 *     The idea here is to simply allow interprocess I/O that exists
 *      only in memory. One process creates the mapping ( FSHM_OWNER )
 *      and several other processes can inherit that mapping,
 *      and share data between them.
 *
 *     There is a real FD associated with this FILE (unlike, open_memstream, for example).
 *
 *     You may choose to use either the buffered variants ( like fwrite, fread),
 *       or the unbuffered variants (like write, read).
 *
 *     Keep in mind that if you use the buffered functions, you must call fflush to
 *       sync the changes.
 *
 *
 *     You can use this object anywhere a FILE* is allowed, or anywhere an fd is allowed
 *      by calling fileno(X) where X is the FILE* returned by this function.
 *
 */
FILE* fshm_open(const char *name, mode_t mode, int fshm_flags);


/**
 * fshm_chgrp - Change the group associated with an existing shmfile stream
 *
 *      You must be the the same user as the FSHM_OWNER of the given stream,
 *        or root in order to change the gid assigned to the shmfile.
 *
 *      The shmfile created by fshm_open is assigned the gid
 *        matching the primary group of the creating user.
 *
 *      The purpose is to allow guests in a different group access rights to
 *        this shmfile ( based on the group bits in the mode set with fshm_open ),
 *        but not just open up access for everybody ( using the "other" bits in mode ).
 *
 *      As non-root, you may change the group to any groups to which you are a member.
 *        This includes your primary gid as well as all supplementry groups to which you belong.
 *        As root, you may change the shmfile's group to any group existing on this system.
 *
 *
 *  fshm_file - A FILE* object as returned by fshm_open.
 *                 You must be the FSHM_OWNER of this shmfile to change permissions.
 *
 *
 *  group     - new gid to replace current group on this shmfile.
 *
 *                  If you are not root, this must be either your primary gid
 *                    or a group contained in your list of supplementry groups.
 *
 *
 *  RETURN VALUE -
 *                   0:  Success
 *                  -1:  Failure (and errno will be set)
 *
 *   See man 2 chown for possible error conditions and values of errno
 * 
 */
int fshm_chgrp(FILE *fshm_file, gid_t group);


/**
 * fshm_chmod - Change the mode on a shmfile
 *
 *   fshm_file - A FILE* object representing an shmfile stream
 *
 *   mode - An octal of permissions. See "chmod" for meaning.
 *
 *  RETURN VALUE -
 *                   0:  Success
 *                  -1:  Failure (and errno will be set)
 *
 *   See man 2 chmod for possible error conditions and values of errno
 */ 
int fshm_chmod(FILE *fshm_file, mode_t mode);



/**
 * fshm_force_destroy - Forcibly destroy the shared memory region associated
 *                       with #name.
 *
 *   Normal operation does not require this, however, if a process opens
 *     a fshm stream as FSHM_OWNER, and does not close it before terminating,
 *     the stream will maintain stuck open without an owner.
 *
 *
 *    name - The name associated with this stream
 *
 *  NOTES - 
 *
 *   The owner should test for this, and forcibly destroy if required.
 *
 *   Like so:
 *
 *   FILE *fObj;
 *
 *   errno = 0;
 *   fObj = fshm_open("/example1", 0775, FSHM_OWNER);
 *
 *   if ( !fObj && errno == EEXIST )
 *   {
 *      fshm_force_destroy("/example1");
 *      fObj = fshm_open("/example1", 0775, FSHM_OWNER);
 *   }
 *
 *
 *  RETURN:
 *    0 on success, -1 on error.
 */
int  fshm_force_destroy(const char *name);

/**
 * shmfile_get_version - Get the version info on this version of shmfile
 *
 *    Sets the value of *major, *minor, *patchlevel, and *extra to the numeric version numbers
 */
void shmfile_get_version(unsigned char *major, unsigned char *minor, unsigned char *patchlevel, const char **extra);


#if defined (__cplusplus)
} /* extern "C" */
#endif

#endif

/* vim: set ts=4 sw=4 st=4 expandtab : */
